tests/functional/test_alias.py

```diff
@@ -0,0 +1,147 @@
+        cmdline = 'my-alias'
+        stdout, _, _ = self.assert_params_for_cmd(cmdline, {})
+        self.assertEqual(len(self.operations_called), 1)
+        self.assertEqual(self.operations_called[0][0].name, 'DescribeRegions')
+        self.assertEqual(stdout.strip('\n'), 'us-east-1')
+
+    def test_subcommand_alias_then_global_params(self):
+        self.add_alias('my-alias', 'ec2 describe-regions')
+        self.parsed_responses = [
+            {
+                'Regions': [
+                    {
+                        'Endpoint': 'ec2.us-east-1.amazonaws.com',
+                        'RegionName': 'us-east-1'
+                    }
+                ]
+            }
+        ]
+        cmdline = 'my-alias '
+        cmdline += '--query=Regions[].RegionName '
+        cmdline += '--output=text'
+        stdout, _, _ = self.assert_params_for_cmd(cmdline, {})
+        self.assertEqual(len(self.operations_called), 1)
+        self.assertEqual(self.operations_called[0][0].name, 'DescribeRegions')
+        self.assertEqual(stdout.strip('\n'), 'us-east-1')
+
+    def test_global_params_then_subcommand_alias(self):
+        self.add_alias('my-alias', 'ec2 describe-regions')
+        self.parsed_responses = [
+            {
+                'Regions': [
+                    {
+                        'Endpoint': 'ec2.us-east-1.amazonaws.com',
+                        'RegionName': 'us-east-1'
+                    }
+                ]
+            }
+        ]
+        cmdline = '--query=Regions[].RegionName '
+        cmdline += '--output=text '
+        cmdline += 'my-alias'
+        stdout, _, _ = self.assert_params_for_cmd(cmdline, {})
+        self.assertEqual(len(self.operations_called), 1)
+        self.assertEqual(self.operations_called[0][0].name, 'DescribeRegions')
+        self.assertEqual(stdout.strip('\n'), 'us-east-1')
+
+    def test_alias_overrides_builtin_command(self):
+        self.add_alias('ec2', 's3api')
+        cmdline = 'ec2 list-buckets'
+        self.assert_params_for_cmd(cmdline, {})
+        self.assertEqual(len(self.operations_called), 1)
+        self.assertEqual(self.operations_called[0][0].name, 'ListBuckets')
+
+    def test_external_alias(self):
+        # The external alias is tested by using mkdir; a command that
+        # is universal for the various OS's we support
+        directory_to_make = os.path.join(self.files.rootdir, 'newdir')
+        self.add_alias('mkdir', '!mkdir %s' % directory_to_make)
+        self.run_cmd('mkdir')
+        self.assertTrue(os.path.isdir(directory_to_make))
+
+    def test_external_alias_then_additonal_args(self):
+        # The external alias is tested by using mkdir; a command that
+        # is universal for the various OS's we support
+        directory_to_make = os.path.join(self.files.rootdir, 'newdir')
+        self.add_alias('mkdir', '!mkdir')
+        self.run_cmd('mkdir %s' % directory_to_make)
+        self.assertTrue(os.path.isdir(directory_to_make))
```

Comment:
It would also be nice to have a posix specific test for the IIFE aliases (as they're one of the more powerful use cases for aliases so it would be good to verify they work).